### PR TITLE
test: generalize public-surface reachability to path-based origin classification (Phase 3)

### PR DESCRIPTION
**Pull request by napetrov**
## Summary

Phase 3 of [`docs/contribute/plans/bug-class-regression-testing.md`](https://github.com/abicheck/abicheck/blob/main/docs/contribute/plans/bug-class-regression-testing.md), generalizing the `policy.public_surface_reachability` bug class (#235 → #834 → #835 → #843).

## Motivation

The plan's Phase 3 invariant: *"a declaration's public/private classification is a function of reachability from an explicit public root through the real include graph, language visibility, and export-table evidence — never directory containment or name-shape alone."*

Two existing seed tests (`test_surface_property.py`, `test_surface_seed_predicate_properties.py`, both added alongside #843) already cover `surface.py`'s reachability-closure/seed-predicate logic well on synthetic snapshots. The still-underserved layer one step below that — `abicheck.provenance`'s path/segment-based `classify_origin`/`_matches_public` classifier, which decides PUBLIC_HEADER vs. PRIVATE_HEADER vs. SYSTEM_HEADER vs. GENERATED for one declaration — had no property/metamorphic coverage of its own, and no test verified that castxml and direct-clang agree on that classification. This PR closes both gaps.

The plan's own text flags Phase 3's full oracle (a generated include-DAG model with `-I`/`-isystem`/`-idirafter` resolution order, `#include` cycles, symlinked roots) as needing its own design review before implementation — so this PR deliberately does not attempt that model, to avoid the under-designed, maintenance-burden outcome the plan itself warns against.

## Changes

- **`tests/test_provenance_classification_properties.py`** (new): a Hypothesis property/metamorphic suite for `classify_origin`/`_matches_public`/`build_public_set`:
  - Metamorphic: classification is invariant to checkout relocation, to `./..` path-spelling noise, and to adding an unrelated public directory.
  - Independent oracle: real directory containment (ground truth known by construction) vs. a sibling-directory string-prefix false positive.
  - A mutant-killing test: the suite is checked against a naive `str.startswith()` reimplementation and demonstrably rejects it.
  - Priority-order tests matching `classify_origin`'s own documented check sequence (public > generated > system).
  - The duplicate-basename accepted trade-off (`_matches_public`'s own documented D3 decision) pinned explicitly, so it can't silently widen or narrow.
- **`tests/test_clang_header_backend_integration.py`**: added `test_clang_and_castxml_agree_on_public_vs_private_header_origin`, extending the existing plain public-surface parity test with cross-backend (castxml vs. direct-clang) agreement on PUBLIC_HEADER/PRIVATE_HEADER origin for a mixed public+private header pair. Skips cleanly when castxml/clang/g++ aren't available (matching this repo's existing marker discipline) — verified locally against the clang-only path (castxml isn't installed in this sandbox); the castxml side is exercised by CI.
- **`tests/regressions/manifest.py`**: registered both new seed tests under `policy.public_surface_reachability`, and refined the `known_gaps` text to describe what this phase added and what it deliberately deferred. Also corrected a prior claim (from PR #885) that AGENTS.md documents a direct-clang path-normalization gap and nested/anonymous-namespace record gaps from #843 — no such text could be found there, so it isn't repeated.

Test-only change; no `abicheck/**/*.py` production code is touched, so no changelog fragment is needed.

## Checklist

- [x] Tests added / updated for new behaviour
- [ ] Changelog fragment added (`scriv create`) if this touches `abicheck/**/*.py` — see `changelog.d/README.md` (N/A — test-only change)
- [ ] Docs updated if user-visible behaviour changed (N/A — no behavior change)
- [x] `pre-commit run --all-files` passes locally (ruff check/format clean on touched files; `tests/regressions/manifest.py` integrity tests pass)
- [x] No new `mypy` or `ruff` errors introduced

---
_Generated by [Claude Code](https://claude.ai/code/session_018USuGdPgyVaPK7z6iB2cKd)_

## Summary by CodeRabbit

- **Tests**
  - Added comprehensive validation for public and private header origin classification, including path variations, directory boundaries, priority rules, and known edge cases.
  - Added cross-backend integration coverage to verify consistent results between Clang and CastXML.
  - Improved test gating so integration checks run only when required tooling is available, while preserving fast test runs elsewhere.
  - Updated regression coverage descriptions to reflect the expanded validation scope and documented exclusions.

### CODE REVIEW

File: tests/test_provenance_classification_properties.py [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+    structure (a public directory, and a header somewhere under or outside
+    it) must classify identically regardless of which absolute checkout
+    root it's rooted at."""
+    assume(root_a != root_b)
+    header_tail = (*pub_rel, *header_rel, f"{filename}.h")
+
+    def classify_under(root: list[str]) -> ScopeOrigin:
+        public_dir = _abspath((*root, *pub_rel))
+        header = _abspath((*root, *header_tail))
+        return _classify(header, [public_dir])
```
[the remark this answers is not in the document]
Comment (napetrov) [reply tests/test_provenance_classification_properties.py]: Confirmed and fixed in 3be473e: every generated `header_tail` did begin with `pub_rel`, so the property only ever compared two `PUBLIC_HEADER` results and couldn't catch a relocation-sensitive implementation. Added an outside-public-dir (private) header case built from a UUID-rooted sibling disjoint from `root`/`pub_rel`/`header_rel`, and now assert each case's expected `ScopeOrigin` explicitly under both checkout roots, not just cross-root equality.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_clang_header_backend_integration.py [hunk 2]
```diff
@@ -268,6 +268,67 @@ def public_funcs(snap: object) -> set[str]:
     assert {e.name for e in clang_snap.enums} == {e.name for e in castxml_snap.enums}
 
 
+def test_clang_and_castxml_agree_on_public_vs_private_header_origin(
+    tmp_path: Path,
+) -> None:
```
[the remark this answers is not in the document]
Comment (napetrov) [reply tests/test_clang_header_backend_integration.py]: Agreed and fixed in 3be473e — added `@pytest.mark.integration` to this test. Applied the same fix to the pre-existing sibling `test_clang_and_castxml_snapshots_agree_on_public_surface`, which had the identical gap for the identical reason (fix the cause, not just this one instance): a host with castxml installed (e.g. a pixi environment, which provisions castxml for the integration/libabigail/abicc marker lanes) would have both castxml-parity tests silently selected and executed by the fast/PR "not integration" lane. Updated the module docstring's note, which previously described the whole module as deliberately unmarked — that blanket statement no longer accurately describes these two specific tests, only the clang-only ones that genuinely need to run on a castxml-absent host.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_provenance_classification_properties.py [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+    root: list[str], pub_rel: list[str], header_rel: list[str], filename: str
+) -> None:
+    """A second, wholly unrelated public directory added to the set must not
+    change how an EXISTING header classifies -- `_matches_public` is a pure
+    `any(...)` over the public-dir list, so an irrelevant addition can only
+    ever add a spurious match, never remove one; the unrelated directory
+    here is built from UUID segments that cannot appear anywhere in the
+    generated header path, so no spurious match is possible either."""
+    public_dir = _abspath((*root, *pub_rel))
+    header = _abspath((*root, *pub_rel, *header_rel, f"{filename}.h"))
```
[the remark this answers is not in the document]
Comment (napetrov) [reply tests/test_provenance_classification_properties.py]: Confirmed and fixed in 7c5dd31 — same-shaped gap as the checkout-relocation property one commit ago. `header` was always nested under `public_dir`, so the property only ever compared two `PUBLIC_HEADER` results before/after the addition, which can't distinguish it from a spurious-match regression. Added a private-header case (a UUID-rooted sibling, disjoint from `root`/`pub_rel`/`header_rel`) checked before and after the addition, asserting `PRIVATE_HEADER` in both cases.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_provenance_classification_properties.py [hunk 4]
```diff
@@ -0,0 +1,411 @@
...
+        "windows kits",
+    }
+)
+_SEGMENT = st.text(
+    alphabet=st.characters(
+        whitelist_categories=("Ll", "Lu", "Nd"), whitelist_characters="_-"
+    ),
+    min_size=1,
+    max_size=8,
+).filter(lambda s: s.lower() not in _RESERVED_TOKENS and s not in (".", ".."))
```
[the remark this answers is not in the document]
Comment (napetrov) [reply tests/test_provenance_classification_properties.py]: Confirmed and fixed in 75c558b — the segment filter only excluded exact reserved tokens, not shapes like `moc_1`/`ui_x`/`x_generated` that `classify_origin`'s generated-header heuristic recognizes once `.h` is appended, which would have made the PRIVATE_HEADER assertions seed-dependent. Added a filter rejecting any segment whose `f"{s}.h"` spelling matches `abicheck.provenance._GENERATED_BASENAME`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_provenance_classification_properties.py [hunk 5]
```diff
@@ -0,0 +1,425 @@
...
+    directory via `_contiguous_subsequence` containment."""
+    assume(root_a != root_b)
+    public_tail = (*pub_rel, *header_rel, f"{filename}.h")
+    private_sibling = (str(uuid.uuid4()), str(uuid.uuid4()))
+    private_tail = (*private_sibling, f"{filename}.h")
+
+    def classify_under(root: list[str], header_tail: tuple[str, ...]) -> ScopeOrigin:
+        public_dir = _abspath((*root, *pub_rel))
+        header = _abspath((*root, *header_tail))
+        return _classify(header, [public_dir])
```
[the remark this answers is not in the document]
Comment (napetrov) [reply tests/test_provenance_classification_properties.py:197]: Confirmed and fixed in 8a670bc. The public directory is now built as a bare relative path (`_abspath(pub_rel)`, no `root` involved at all), independent of which root the header relocates to — matching `provenance.py`'s own docstring rationale (build-recorded absolute paths bearing no resemblance to the user-described public set). Added disjointness assumptions between `pub_rel` and `root`/`header_rel` to avoid an accidental segment collision producing a false containment match, since decoupling the root removed the previous shared-prefix guarantee. Verified empirically against a hand-written root-coupled-prefix mutant matching the shape you described: it now fails the fixed test (returns `PRIVATE_HEADER` where the real `classify_origin` gives `PUBLIC_HEADER`), confirming the fix actually detects it.

---
_Generated by [Claude Code](https://claude.ai/code)_

[the remark this answers is not in the document]
Comment (napetrov) [reply tests/test_provenance_classification_properties.py:197]: Investigated this one rather than applying it — I believe the suggestion is incorrect for how this codebase's `ScopeOrigin`/`classify_origin` is actually designed, and no change was made.

`provenance.py`'s own module docstring states its classification is deliberately path/segment-based and independent of export status — that's not an implementation gap, it's the documented contract. Verified this is intentional, not accidental, by checking what consumes `ScopeOrigin.PRIVATE_HEADER`:

- `abicheck/checker_policy.py` has a dedicated `ChangeKind.PRIVATE_HEADER_LEAK = "private_header_leak"` — "a public header pulls in a private/non-installed header → RISK".
- `abicheck/surface.py`'s `REASON_PRIVATE_HEADER = "decl originates in a non-public project header"` is a *provenance-driven* demotion reason, explicitly kept orthogonal to the *separate* `REASON_NOT_EXPORTED`/`REASON_NON_PUBLIC_TYPE` reachability/linkage reasons (see that module's own comment: "``private-header``/``system-header`` are provenance-driven ... ``not-exported``/``non-public-type`` are the linkage/reachability reasons").
- `Function.visibility` (computed from real ELF export-table evidence in `dumper_clang.py`'s `_visibility()`/`dumper_castxml.py`) is the *separate* field that does carry export/reachability information — `Function.origin` is not meant to duplicate it.

So a declaration that is genuinely reachable via `#include` from a public root, genuinely exported by the linker, but declared in a header the user did not name as public, is *exactly* the case `ScopeOrigin.PRIVATE_HEADER` + `Visibility.PUBLIC` together are designed to represent — it's the signal the `private_header_leak` detector and `surface.py`'s demotion logic consume, not a defect to fix by promoting the origin. Making `classify_origin` incorporate include-graph reachability and export evidence, as suggested, would collapse this distinction and break that detection. `detail_helper`'s asserted `PRIVATE_HEADER` origin in this test is the correct, intentional value; the Phase 3 invariant's "reachability... and export-table evidence" language describes `surface.py`'s reachability closure (already covered by the pre-existing `test_surface_property.py`/`test_surface_seed_predicate_properties.py` seed tests for this bug class), not `provenance.py`'s narrower, deliberately origin-only classifier this test targets.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_clang_castxml_origin_parity.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+
+    clang_origins = origins(clang_snap)
+    castxml_origins = origins(castxml_snap)
+    # Both backends must actually see both functions -- an empty/partial
+    # dict would make the equality below vacuously true.
+    assert set(clang_origins) == {"api_call", "detail_helper"}
+    assert set(castxml_origins) == {"api_call", "detail_helper"}
+    assert clang_origins == castxml_origins
+    assert clang_origins["api_call"] == ScopeOrigin.PUBLIC_HEADER.value
+    assert clang_origins["detail_helper"] == ScopeOrigin.PRIVATE_HEADER.value
```
[the remark this answers is not in the document]
Comment (napetrov) [reply tests/test_clang_castxml_origin_parity.py:125]: (Apologies — my rebuttal for this comment was accidentally posted onto the "Decouple the public root" thread above instead of here; repeating it on the correct thread.)

Investigated this rather than applying it — I believe the suggestion is incorrect for how this codebase's `ScopeOrigin`/`classify_origin` is actually designed, and no change was made.

`provenance.py`'s own module docstring states its classification is deliberately path/segment-based and independent of export status — that's not an implementation gap, it's the documented contract. Confirmed this is intentional, not accidental, by checking what consumes `ScopeOrigin.PRIVATE_HEADER`:

- `abicheck/checker_policy.py` has a dedicated `ChangeKind.PRIVATE_HEADER_LEAK = "private_header_leak"` — "a public header pulls in a private/non-installed header → RISK".
- `abicheck/surface.py`'s `REASON_PRIVATE_HEADER = "decl originates in a non-public project header"` is a *provenance-driven* demotion reason, explicitly kept orthogonal to the *separate* `REASON_NOT_EXPORTED`/`REASON_NON_PUBLIC_TYPE` reachability/linkage reasons (that module's own comment: "``private-header``/``system-header`` are provenance-driven ... ``not-exported``/``non-public-type`` are the linkage/reachability reasons").
- `Function.visibility` (computed from real ELF export-table evidence in `dumper_clang.py`'s `_visibility()`/`dumper_castxml.py`) is the field that carries export/reachability information — `Function.origin` doesn't duplicate it. Verified directly: in this test's own scenario, `detail_helper` and `api_call` both come back with the correct, real `Visibility` computed from the binary's actual dynamic symbol table (independent of `origin`).

So a declaration genuinely reachable via `#include` from a public root, genuinely exported by the linker, but declared in a header the user didn't name as public, is *exactly* the case `ScopeOrigin.PRIVATE_HEADER` + an independently-computed `Visibility` are designed to represent together — it's the signal `private_header_leak` and `surface.py`'s demotion logic consume, not a defect to fix by promoting the origin. Making `classify_origin` incorporate include-graph reachability and export evidence, as suggested, would collapse that distinction and break that detection. `detail_helper`'s asserted `PRIVATE_HEADER` origin here is the correct, intentional value; the Phase 3 invariant's "reachability... and export-table evidence" language describes `surface.py`'s reachability closure (already covered by this bug class's pre-existing `test_surface_property.py`/`test_surface_seed_predicate_properties.py` seed tests), not `provenance.py`'s narrower, deliberately origin-only classifier this test targets.

---
_Generated by [Claude Code](https://claude.ai/code)_
